File: front/central/src/services/modules/torneo/domain/ports.ts
import {
    ActionResponse,
    CreateEquipoDTO,
    CreateEventoDTO,
    CreateJugadorDTO,
    CreatePartidoDTO,
    CreateTorneoDTO,
    Equipo,
    Evento,
    Goleador,
    GetJugadoresParams,
    GetPartidosParams,
    Jugador,
    PaginatedResponse,
    PaginationParams,
    Partido,
    PosicionTabla,
    ResultadoDTO,
    SingleResponse,
    Torneo,
    UpdateEquipoDTO,
    UpdateJugadorDTO,
    UpdatePartidoDTO,
    UpdateTorneoDTO,
} from './types';

export interface ITorneoRepository {
    // Torneos (scope por negocio)
    getTorneos(businessId: number, params?: PaginationParams): Promise<PaginatedResponse<Torneo>>;
    getTorneo(id: number): Promise<SingleResponse<Torneo>>;
    createTorneo(businessId: number, data: CreateTorneoDTO): Promise<SingleResponse<Torneo>>;
    updateTorneo(id: number, data: UpdateTorneoDTO): Promise<SingleResponse<Torneo>>;
    deleteTorneo(id: number): Promise<ActionResponse>;

    // Equipos
    getEquipos(torneoId: number, params?: PaginationParams): Promise<PaginatedResponse<Equipo>>;
    getEquipoById(torneoId: number, id: number): Promise<SingleResponse<Equipo>>;
    createEquipo(torneoId: number, data: CreateEquipoDTO): Promise<SingleResponse<Equipo>>;
    updateEquipo(torneoId: number, id: number, data: UpdateEquipoDTO): Promise<SingleResponse<Equipo>>;
    deleteEquipo(torneoId: number, id: number): Promise<ActionResponse>;

    // Jugadores
    getJugadores(torneoId: number, params?: GetJugadoresParams): Promise<PaginatedResponse<Jugador>>;
    getJugadorById(torneoId: number, id: number): Promise<SingleResponse<Jugador>>;
    createJugador(torneoId: number, data: CreateJugadorDTO): Promise<SingleResponse<Jugador>>;
    updateJugador(torneoId: number, id: number, data: UpdateJugadorDTO): Promise<SingleResponse<Jugador>>;
    deleteJugador(torneoId: number, id: number): Promise<ActionResponse>;

    // Partidos
    getPartidos(torneoId: number, params?: GetPartidosParams): Promise<PaginatedResponse<Partido>>;
    getPartidoById(torneoId: number, id: number): Promise<SingleResponse<Partido>>;
    createPartido(torneoId: number, data: CreatePartidoDTO): Promise<SingleResponse<Partido>>;
    updatePartido(torneoId: number, id: number, data: UpdatePartidoDTO): Promise<SingleResponse<Partido>>;
    deletePartido(torneoId: number, id: number): Promise<ActionResponse>;
    registrarResultado(torneoId: number, id: number, data: ResultadoDTO): Promise<SingleResponse<Partido>>;

    // Eventos
    getEventos(torneoId: number, partidoId: number): Promise<SingleResponse<Evento[]>>;
    createEvento(torneoId: number, partidoId: number, data: CreateEventoDTO): Promise<SingleResponse<Evento>>;
    deleteEvento(torneoId: number, id: number): Promise<ActionResponse>;

    // Tabla / Goleadores
    getTabla(torneoId: number): Promise<SingleResponse<PosicionTabla[]>>;
    getGoleadores(torneoId: number): Promise<SingleResponse<Goleador[]>>;

    // Fixture
    generarFixture(torneoId: number): Promise<ActionResponse>;
}

File: back/central/services/modules/torneo/internal/domain/ports.go
package domain

import "context"

// ITorneoRepository define el puerto de persistencia del módulo de torneo.
type ITorneoRepository interface {
	// Torneos
	CreateTorneo(ctx context.Context, dto CreateTorneoDTO) (*Torneo, error)
	UpdateTorneo(ctx context.Context, id uint, dto UpdateTorneoDTO) (*Torneo, error)
	DeleteTorneo(ctx context.Context, id uint) error
	GetTorneoByID(ctx context.Context, id uint) (*Torneo, error)
	GetTorneos(ctx context.Context, businessID uint, p PaginationParams) ([]Torneo, int, error)

	// Equipos
	CreateEquipo(ctx context.Context, dto CreateEquipoDTO) (*Equipo, error)
	UpdateEquipo(ctx context.Context, torneoID, id uint, dto UpdateEquipoDTO) (*Equipo, error)
	DeleteEquipo(ctx context.Context, torneoID, id uint) error
	GetEquipoByID(ctx context.Context, torneoID, id uint) (*Equipo, error)
	GetEquipos(ctx context.Context, torneoID uint, p PaginationParams) ([]Equipo, int, error)
	ListEquipos(ctx context.Context, torneoID uint) ([]Equipo, error)

	// Jugadores
	CreateJugador(ctx context.Context, dto CreateJugadorDTO) (*Jugador, error)
	UpdateJugador(ctx context.Context, torneoID, id uint, dto UpdateJugadorDTO) (*Jugador, error)
	DeleteJugador(ctx context.Context, torneoID, id uint) error
	GetJugadorByID(ctx context.Context, torneoID, id uint) (*Jugador, error)
	GetJugadores(ctx context.Context, filters JugadorFilters, p PaginationParams) ([]Jugador, int, error)

	// Partidos
	CreatePartido(ctx context.Context, dto CreatePartidoDTO) (*Partido, error)
	UpdatePartido(ctx context.Context, torneoID, id uint, dto UpdatePartidoDTO) (*Partido, error)
	DeletePartido(ctx context.Context, torneoID, id uint) error
	GetPartidoByID(ctx context.Context, torneoID, id uint) (*Partido, error)
	GetPartidos(ctx context.Context, filters PartidoFilters, p PaginationParams) ([]Partido, int, error)
	SetResultado(ctx context.Context, torneoID, id uint, dto ResultadoPartidoDTO) (*Partido, error)
	DeletePartidosNoJugados(ctx context.Context, torneoID uint) error
	CountPartidosJugados(ctx context.Context, torneoID uint) (int, error)
	CreatePartidos(ctx context.Context, partidos []CreatePartidoDTO) ([]Partido, error)

	// Eventos
	CreateEvento(ctx context.Context, dto CreateEventoDTO) (*PartidoEvento, error)
	DeleteEvento(ctx context.Context, id uint) error
	GetEventoByID(ctx context.Context, id uint) (*PartidoEvento, error)
	GetEventosByPartido(ctx context.Context, partidoID uint) ([]PartidoEvento, error)

	// Calculados
	GetTablaPosiciones(ctx context.Context, torneoID uint) ([]TablaPosicion, error)
	GetGoleadores(ctx context.Context, torneoID uint) ([]Goleador, error)
}
